### PR TITLE
CryoDRGN2

### PR DESCRIPTION
**Pull request by adamlerer**
Release of [cryoDRGN2](https://openaccess.thecvf.com/content/ICCV2021/html/Zhong_CryoDRGN2_Ab_Initio_Neural_Reconstruction_of_3D_Protein_Structures_From_ICCV_2021_paper.html)

Available via `cryodrgn abinit_het ...` . Instructions added to the README.

### CODE REVIEW

File: cryodrgn/dataset.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def __init__(self, mrcfile, norm=None, keepreal=False, invert_data=False, ind=No
         assert ny == nx, "Images must be square"
         assert ny % 2 == 0, "Image size must be even. Is this a preprocessed dataset? Use the --preprocessed flag if so."
         log('Loaded {} {}x{} images'.format(N, ny, nx))
-        self.particles = particles
+        self.particles = particles.astype(np.float32)
```
Comment (adamlerer) [inline cryodrgn/dataset.py:55]: @zhonge is this correct?

File: cryodrgn/make_healpy.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import numpy as np
```
Comment (adamlerer) [inline cryodrgn/make_healpy.py:1]: Where does this belong?

File: cryodrgn/models.py [hunk 3]
```diff
@@ -705,21 +706,42 @@ def forward(self, x, x_tilt):
...
+    # print('half', input.shape, weight.shape)
+    return F.linear(input, weight.half(), bias.half())
+
+def single_linear(input, weight, bias):
+    # print('single', input.shape, weight.shape)
+    # assert input.shape[0] < 10000
+
+    return F.linear(input, weight, bias)
+
+class MyLinear(nn.Linear):
```
Comment (adamlerer) [inline cryodrgn/models.py:733]: what on earth is this? I have no recollection.

Comment (adamlerer) [reply cryodrgn/models.py:733]: Okay, it's from a (partial?) attempt to implement half-precision. We did run half-precision experiments for CD2, so it probably works at least in that context. No guarantee that it works everywhere though.

**Review by zhonge — APPROVED** [review]

**Pull request merged**